File: src/main/java/com/learning/leetcode/arrays/easy/PlusOne.java
package com.learning.leetcode.arrays.easy;

import java.util.Arrays;

public class PlusOne {
    public static void main(String[] args) {
        int[] digits = {9,8,7,6,5,4,3,2,1,0};
        System.out.println("Original array" + Arrays.toString(digits));
        System.out.println("Plus one array" + Arrays.toString(plusOne(digits)));
        digits = new int[]{9};
        System.out.println("Original array" + Arrays.toString(digits));
        System.out.println("Plus one array" + Arrays.toString(plusOne(digits)));
        digits = new int[]{9,9};
        System.out.println("Original array" + Arrays.toString(digits));
        System.out.println("Plus one array" + Arrays.toString(plusOne(digits)));

    }

    private static int[] plusOne(int[] digits){
        for (int i = digits.length - 1; i >= 0; i--) {
            if (digits[i] < 9) {
                digits[i]++;
                return digits;
            }
            digits[i] = 0;
        }
        System.out.println(Arrays.toString(digits));
        // This is only for the scenarios like [9],[9,9],[9,9,9] etc.
        digits = new int[digits.length + 1];
        digits[0] = 1;
        return digits;
    }
}

File: src/main/java/com/learning/leetcode/arrays/easy/PlusOne.py
from typing import List


def plusOne(digits: List[int]) -> List[int]:
    for i in range(len(digits)-1, - 1, -1):
        if digits[i] < 9:
            digits[i] += 1
            return digits
        # Set current value to 0 if 9 and
        # increment digits[i] in the next iteration
        digits[i] = 0
    digits = [0] * (len(digits) + 1)
    digits[0] = 1
    return digits


def main():
    digits = [9, 8, 7, 6, 5, 9]
    print('Given -> %20s' % digits)
    print('Plus one  -> %16s' % plusOne(digits))
    digits = [9]
    print('Given -> %20s' % digits)
    print('Plus one  -> %16s' % plusOne(digits))
    digits = [9, 9]
    print('Given -> %20s' % digits)
    print('Plus one  -> %16s' % plusOne(digits))


main()
